### PR TITLE
[stdlib] Improve the mangled names of some @_alwaysEmitIntoClient functions

### PR DESCRIPTION
**Pull request by glessard**
Some mangled names for non-abi functions included `$` by mistake.
This PR also improves the explanation for compatibility entry points that use the mangled names of newer, non-abi functions.

No related issue.

### CODE REVIEW

**Comment by glessard** [comment]
@swift-ci please test

**Review by stephentyrone — APPROVED** [review]

File: stdlib/public/core/UnsafeRawBufferPointer.swift.gyb [hunk 1]
```diff
@@ -479,7 +480,7 @@ extension Unsafe${Mutable}RawBufferPointer {
 
   // This unavailable implementation uses the expected mangled name
   // of `storeBytes<T>(of:toByteOffset:as:)`, and provides an entry point for
-  // any binary compiled against the stlib binary for Swift 5.6 and older.
+  // any binary linked against the stlib binary for Swift 5.6 and older.
```
Comment (stephentyrone) [inline stdlib/public/core/UnsafeRawBufferPointer.swift.gyb:483]: stdlib

File: stdlib/public/core/UnsafePointer.swift [hunk 2]
```diff
@@ -1032,8 +1033,9 @@ public struct UnsafeMutablePointer<Pointee>: _Pointer {
   }
 
   // This unavailable implementation uses the expected mangled name
-  // of `withMemoryRebound`, and provides an entry point for any
-  // binary compiled against the stlib binary for Swift 5.6 and older.
+  // of `withMemoryRebound<T, Result>(to:capacity:_:)`, and provides
+  // an entry point for any binary linked against the stlib binary
```
Comment (stephentyrone) [inline stdlib/public/core/UnsafePointer.swift:1037]: stdlib

File: stdlib/public/core/UnsafePointer.swift [hunk 3]
```diff
@@ -326,8 +326,9 @@ public struct UnsafePointer<Pointee>: _Pointer {
...
```
Comment (stephentyrone) [inline stdlib/public/core/UnsafePointer.swift:330]: stdlib

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb [hunk 4]
```diff
@@ -763,8 +763,9 @@ extension Unsafe${Mutable}BufferPointer {
   }
 
   // This unavailable implementation uses the expected mangled name
-  // of `withMemoryRebound`, and provides an entry point for any
-  // binary compiled against the stlib binary for Swift 5.6 and older.
+  // of `withMemoryRebound<T, Result>(to:_:)`, and provides
+  // an entry point for any binary linked against the stlib binary
```
Comment (stephentyrone) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb:767]: stdlib

**Comment by glessard** [comment]
@stephentyrone, typo hunter, strikes again. Thanks!

**Comment by glessard** [comment]
@swift-ci please test and merge

**Pull request merged**